### ISSUE TITLE
Feature request: runtime evaluation of docker tarball on DNANexus

### ISSUE THREAD

**Issue by zhoulytwinyu** [dnanexus/dxCompiler#119]
# Background:
Currently, tarball docker images gets evaluated at compile time and would error out if the docker image is not found. This is true when the tarball is declared using either dx path syntax ("dx://GenomeSequenceProject:/A/B/myOrgTools") or dx file id syntax (e.g. "dx://file-xxxx"). I will be focusing on **dx path syntax** for this feature request.
Once the wdl is compiled with tarball docker images, the workflow become statically associated with the images. If the images are swapped with tarballs by the same name, the workflow still tries to load the old tarball images. To use new images, the wdl files need to be re-compiled.

One way to trick DNANexus, is to add a constant variable in docker image, like below. This way, DNANexus will try to locate the image at runtime, instead of compile time.
```
task  ... {
  String tag='1.2.3'

  runtime {
    docker: "dx://GenomeSequenceProject:/A/B/myOrgTools-~{tag}.tar"
  }
  ...
}
```

# Request:
If the user specifies docker images using dx path syntax, dxCompiler should skip checking the availability of the images. Instead, it should check them at runtime.
We could do the same for docker images specified using dx file syntac, given the similarity between `docker load ${file_id}` vs `docker pull docker.io/${layer_id}`. But I digress.

# Reason:
1. CI/CD pipeline
Our wdl files and docker files are kept in two separate git repositories and have/will have separate CI/CD pipelines.
The wdl file repo's CI/CD compiles wdl to DX.
The docker file repo's CI/CD builds and pushes images to dockerhub as well as save to DX.
The compile time check would require us to run the CI/CD in a specific order. Or in some god-forbid cases when docker images are recompiled, the affected wdl needs to be re-compiled.
Having tarball docker images to be evaluated at runtime mitigate this problem. wdl and docker file can be managed by separate CI/CD, just like when we stick to dockerhub only deployment.

2. Similarity with `docker pull`
If an image is specified using a registry url, the image is pulled at runtime and not checked at compile time. Similarly, I think it makes more sense to `docker load` tarball images at runtime and not check them at compile time.
It gives us a freedom to swap the images without needing to recompile the workflow. This is a bad practice, I know, but it is convenient during development.

# TL; DR
Having tarball docker images evaluated at runtime helps CI/CD.

# Other
My request aside, I'd love to here your suggestions on a proper CI/CD solution that works with the current dxCompiler implementation as well.

Best
Joe

**Comment by jdidion** [comment]
Thanks for this suggestion. I think it is doable, although we a compiler flag may be used to enable it. I'll add it to our internal feature request tracker.

**Comment by commandlinegirl** [comment]
We’ve considered this feature and while is useful in certain cases it makes it hard to keep the guarantee of reproducibility on the DNAnexus platform (as every job run would depend on a different execution environment in a Docker container) and we’ve decided not to implement it. Thank you for submitting the issue and let us know if you have any other questions.
